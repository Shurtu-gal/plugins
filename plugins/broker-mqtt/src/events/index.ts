export * from "./createServerDotEnv";
export * from "./createPackageJson";
export * from "./createAppModule";
export * from "./createDockerCompose";
