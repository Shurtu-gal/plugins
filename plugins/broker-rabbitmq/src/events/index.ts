export * from "./createServer";
export * from "./createMessageBrokerService";
export * from "./createConnectMicroservices";
