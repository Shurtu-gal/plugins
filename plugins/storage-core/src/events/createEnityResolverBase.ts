import {
  CreateEntityResolverBaseParams,
  DsgContext,
  EntityField,
  EnumDataType,
} from "@amplication/code-gen-types";
import {
  addImports,
  getClassDeclarationById,
  importNames,
  interpolate,
} from "../util/ast";
import { namedTypes, builders } from "ast-types";
import { pascalCase } from "pascal-case";

export const beforeCreateEntityResolverBase = async (
  context: DsgContext,
  eventParams: CreateEntityResolverBaseParams,
) => {
  const { entity, template, templateMapping, resolverBaseId } = eventParams;
  const fileDataTypeFields = entity.fields.filter(
    (field) => field.dataType === EnumDataType.File,
  );

  if (fileDataTypeFields.length === 0) {
    return eventParams;
  }

  interpolate(template, templateMapping);

  const classDeclaration = getClassDeclarationById(template, resolverBaseId);

  fileDataTypeFields.forEach((field) =>
    addFileResolverMethods(context, classDeclaration, eventParams, field),
  );

  const graphQLUploadImport = importNames(
    [builders.identifier("GraphQLUpload")],
    "graphql-upload",
  );

  const fileUploadImport = importNames(
    [builders.identifier("FileUpload")],
    "src/storage/base/storage.types",
  );

  const imports = [graphQLUploadImport, fileUploadImport];

  addImports(template, imports);

  return eventParams;
};

const addFileResolverMethods = (
  context: DsgContext,
  classDeclaration: namedTypes.ClassDeclaration,
  eventParams: CreateEntityResolverBaseParams,
  field: EntityField,
) => {
  const { name } = field;
  const { entity } = eventParams;

  const entityId = builders.identifier(pascalCase(entity.name));
  const uploadMethodId = builders.identifier(`upload${pascalCase(name)}`);
  const deleteMethodId = builders.identifier(`delete${pascalCase(name)}`);

  //@graphql.Mutation(() => User)
  const mutationDecorator = builders.decorator(
    builders.callExpression(builders.identifier("graphql.Mutation"), [
      builders.arrowFunctionExpression([], entityId),
    ]),
  );

  const argsParameter = builders.identifier("args");
  argsParameter.typeAnnotation = builders.tsTypeAnnotation(
    builders.tsTypeReference(
      builders.identifier(`${pascalCase(entity.name)}FindUniqueArgs`),
    ),
  );
  // @ts-expect-error decorators are not supported by ast-types
  argsParameter.decorators = [
    builders.decorator(
      builders.callExpression(builders.identifier("graphql.Args"), []),
    ),
  ];

  const fileParameter = builders.identifier("file");
  fileParameter.typeAnnotation = builders.tsTypeAnnotation(
    builders.tsTypeReference(builders.identifier("FileUpload")),
  );
  // @ts-expect-error decorators are not supported by ast-types
  fileParameter.decorators = [
    builders.decorator(
      builders.callExpression(builders.identifier("graphql.Args"), [
        builders.objectExpression([
          builders.objectProperty(
            builders.identifier("name"),
            builders.stringLiteral("file"),
          ),
          builders.objectProperty(
            builders.identifier("type"),
            builders.arrowFunctionExpression(
              [],
              builders.identifier("GraphQLUpload"),
            ),
          ),
        ]),
      ]),
    ),
  ];

  const uploadMethod = getUploadMethod(
    [fileParameter, argsParameter],
    entityId,
    mutationDecorator,
    uploadMethodId,
    name,
  );

  const deleteMethod = getDeleteMethod(
    [argsParameter],
    entityId,
    mutationDecorator,
    deleteMethodId,
    name,
  );

  classDeclaration.body.body.push(uploadMethod, deleteMethod);
};

const getUploadMethod = (
  parameters: namedTypes.Identifier[],
  entityId: namedTypes.Identifier,
  decorator: namedTypes.Decorator,
  methodId: namedTypes.Identifier,
  fieldName: string,
) => {
  const method = builders.classMethod.from({
    kind: "method",
    async: true,
    key: methodId,
    params: parameters,
    decorators: [decorator],
    returnType: builders.tsTypeAnnotation(
      builders.tsTypeReference(
        builders.identifier("Promise"),
        builders.tsTypeParameterInstantiation([
          builders.tsTypeReference(entityId),
        ]),
      ),
    ),
    body: builders.blockStatement([
      builders.returnStatement(
        builders.awaitExpression(
          builders.callExpression(
            builders.memberExpression(
              builders.memberExpression(
                builders.thisExpression(),
                builders.identifier("service"),
              ),
              builders.identifier(`upload${pascalCase(fieldName)}`),
            ),
            [builders.identifier("args"), builders.identifier("file")],
          ),
        ),
      ),
    ]),
  });

  return method;
};

const getDeleteMethod = (
  parameters: namedTypes.Identifier[],
  entityId: namedTypes.Identifier,
  decorator: namedTypes.Decorator,
  methodId: namedTypes.Identifier,
  fieldName: string,
) => {
  const method = builders.classMethod.from({
    kind: "method",
    key: methodId,
    async: true,
    params: parameters,
    decorators: [decorator],
    returnType: builders.tsTypeAnnotation(
      builders.tsTypeReference(
        builders.identifier("Promise"),
        builders.tsTypeParameterInstantiation([
          builders.tsTypeReference(entityId),
        ]),
      ),
    ),
    body: builders.blockStatement([
      builders.returnStatement(
        builders.awaitExpression(
          builders.callExpression(
            builders.memberExpression(
              builders.memberExpression(
                builders.thisExpression(),
                builders.identifier("service"),
              ),
              builders.identifier(`delete${pascalCase(fieldName)}`),
            ),
            [builders.identifier("args")],
          ),
        ),
      ),
    ]),
  });

  return method;
};
