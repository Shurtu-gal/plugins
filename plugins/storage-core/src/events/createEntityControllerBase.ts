import {
  CreateEntityControllerBaseParams,
  DsgContext,
  EntityField,
  EnumDataType,
} from "@amplication/code-gen-types";
import {
  addImports,
  getClassDeclarationById,
  importNames,
  interpolate,
} from "../util/ast";
import { builders, namedTypes } from "ast-types";
import { pascalCase } from "pascal-case";

export const beforeCreateEntityControllerBase = async (
  context: DsgContext,
  eventParams: CreateEntityControllerBaseParams,
) => {
  const { controllerBaseId, entity, template, templateMapping } = eventParams;
  const fileDataTypeFields = entity.fields.filter(
    (field) => field.dataType === EnumDataType.File,
  );

  if (fileDataTypeFields.length === 0) {
    return eventParams;
  }

  interpolate(template, templateMapping);

  const classDeclaration = getClassDeclarationById(template, controllerBaseId);

  fileDataTypeFields.forEach((field) =>
    addFileFieldMethods(context, classDeclaration, eventParams, field),
  );

  const fileInterceptorImport = importNames(
    [builders.identifier("FileInterceptor")],
    "@nestjs/platform-express",
  );

  const ResponseImport = importNames(
    [builders.identifier("Response")],
    "express",
  );

  addImports(template, [fileInterceptorImport, ResponseImport]);

  return eventParams;
};

const addFileFieldMethods = async (
  context: DsgContext,
  classDeclaration: namedTypes.ClassDeclaration,
  eventParams: CreateEntityControllerBaseParams,
  field: EntityField,
) => {
  const { name } = field;
  const { entity } = eventParams;

  const entityId = builders.identifier(pascalCase(entity.name));
  const uploadMethodId = builders.identifier(`upload${pascalCase(name)}`);
  const downloadMethodId = builders.identifier(`download${pascalCase(name)}`);
  const deleteMethodId = builders.identifier(`delete${pascalCase(name)}`);

  // @common.Param() params: UserWhereUniqueInput
  const paramParameter = builders.identifier("params");
  paramParameter.typeAnnotation = builders.tsTypeAnnotation(
    builders.tsTypeReference(
      builders.identifier(`${pascalCase(entity.name)}WhereUniqueInput`),
    ),
  );
  // @ts-expect-error - property is missing in ast-types
  paramParameter.decorators = [
    builders.decorator(
      builders.callExpression(builders.identifier("common.Param"), []),
    ),
  ];

  // @common.UploadedFile() file: Express.Multer.File
  const fileParameter = builders.identifier("file");
  fileParameter.typeAnnotation = builders.tsTypeAnnotation(
    builders.tsTypeReference(builders.identifier("Express.Multer.File")),
  );
  // @ts-expect-error - property is missing in ast-types
  fileParameter.decorators = [
    builders.decorator(
      builders.callExpression(builders.identifier("common.UploadedFile"), []),
    ),
  ];

  // @common.Res({ passthrough: true }) res: Response
  const resParameter = builders.identifier("res");
  resParameter.typeAnnotation = builders.tsTypeAnnotation(
    builders.tsTypeReference(builders.identifier("Response")),
  );
  // @ts-expect-error - property is missing in ast-types
  resParameter.decorators = [
    builders.decorator(
      builders.callExpression(builders.identifier("common.Res"), [
        builders.objectExpression([
          builders.property(
            "init",
            builders.identifier("passthrough"),
            builders.literal(true),
          ),
        ]),
      ]),
    ),
  ];

  const uploadMethod = getUploadMethod(
    [paramParameter, fileParameter],
    uploadMethodId,
    entityId,
    name,
  );

  const downloadMethod = getDownloadMethod(
    [paramParameter, resParameter],
    downloadMethodId,
    entityId,
    name,
  );

  const deleteMethod = getDeleteMethod(
    [paramParameter],
    deleteMethodId,
    entityId,
    name,
  );

  classDeclaration.body.body.push(uploadMethod, downloadMethod, deleteMethod);
};

const swaggerApiNotFoundResponseDecorator = builders.decorator(
  builders.callExpression(builders.identifier("swagger.ApiNotFoundResponse"), [
    builders.objectExpression([
      builders.property(
        "init",
        builders.identifier("type"),
        builders.memberExpression(
          builders.identifier("errors"),
          builders.identifier("NotFoundException"),
        ),
      ),
    ]),
  ]),
);

const getDownloadMethod = (
  params: namedTypes.Identifier[],
  methodId: namedTypes.Identifier,
  entityId: namedTypes.Identifier,
  name: string,
) => {
  const getDecorator = builders.decorator(
    builders.callExpression(builders.identifier("common.Get"), [
      builders.literal(`:id/${name}`),
    ]),
  );

  const swaggerApiParamDecorator = builders.decorator(
    builders.callExpression(builders.identifier("swagger.ApiParam"), [
      builders.objectExpression([
        builders.property(
          "init",
          builders.identifier("name"),
          builders.literal("id"),
        ),
        builders.property(
          "init",
          builders.identifier("type"),
          builders.literal("string"),
        ),
        builders.property(
          "init",
          builders.identifier("required"),
          builders.literal(true),
        ),
      ]),
    ]),
  );

  const swaggerApiOkResponseDecorator = builders.decorator(
    builders.callExpression(builders.identifier("swagger.ApiOkResponse"), [
      builders.objectExpression([
        builders.property(
          "init",
          builders.identifier("type"),
          builders.identifier("common.StreamableFile"),
        ),
      ]),
    ]),
  );

  return builders.classMethod.from({
    async: true,
    key: methodId,
    params: params,
    decorators: [
      getDecorator,
      swaggerApiParamDecorator,
      swaggerApiOkResponseDecorator,
      swaggerApiNotFoundResponseDecorator,
    ],
    returnType: builders.tsTypeAnnotation(
      builders.tsTypeReference(
        builders.identifier("Promise"),
        builders.tsTypeParameterInstantiation([
          builders.tsTypeReference(
            builders.identifier("common.StreamableFile"),
          ),
        ]),
      ),
    ),
    body: builders.blockStatement([
      builders.variableDeclaration("const", [
        builders.variableDeclarator(
          builders.identifier("result"),
          builders.awaitExpression(
            builders.callExpression(
              builders.memberExpression(
                builders.identifier("this.service"),
                builders.identifier(`download${pascalCase(name)}`),
              ),
              [
                builders.objectExpression([
                  builders.property(
                    "init",
                    builders.identifier("where"),
                    builders.identifier("params"),
                  ),
                ]),
              ],
            ),
          ),
        ),
      ]),
      builders.ifStatement(
        builders.binaryExpression(
          "===",
          builders.identifier("result"),
          builders.literal(null),
        ),
        builders.blockStatement([
          builders.throwStatement(
            builders.newExpression(
              builders.identifier("errors.NotFoundException"),
              [
                builders.templateLiteral(
                  [
                    builders.templateElement(
                      {
                        raw: `No resource was found for `,
                        cooked: `No resource was found for `,
                      },
                      false,
                    ),
                  ],
                  [],
                ),
                builders.callExpression(builders.identifier("JSON.stringify"), [
                  builders.identifier("params"),
                ]),
              ],
            ),
          ),
        ]),
      ),
      builders.expressionStatement(
        builders.callExpression(
          builders.memberExpression(
            builders.identifier("res"),
            builders.identifier("setHeader"),
          ),
          [
            builders.literal("Content-Disposition"),
            builders.templateLiteral(
              [
                builders.templateElement(
                  {
                    raw: `attachment; filename=`,
                    cooked: `attachment; filename=`,
                  },
                  false,
                ),
              ],
              [builders.identifier("result.filename")],
            ),
          ],
        ),
      ),
      builders.expressionStatement(
        builders.callExpression(
          builders.memberExpression(
            builders.identifier("res"),
            builders.identifier("setHeader"),
          ),
          [
            builders.literal("Content-Type"),
            builders.identifier("result.mimetype"),
          ],
        ),
      ),
      builders.returnStatement(builders.identifier("result.stream")),
    ]),
  });
};

const getUploadMethod = (
  params: namedTypes.Identifier[],
  methodId: namedTypes.Identifier,
  entityId: namedTypes.Identifier,
  name: string,
) => {
  const putDecorator = builders.decorator(
    builders.callExpression(builders.identifier("common.Put"), [
      builders.literal(`:id/${name}`),
    ]),
  );

  const interceptorDecorator = builders.decorator(
    builders.callExpression(builders.identifier("common.UseInterceptors"), [
      builders.callExpression(builders.identifier("FileInterceptor"), [
        builders.literal("file"),
      ]),
    ]),
  );

  const consumesDecorator = builders.decorator(
    builders.callExpression(builders.identifier("swagger.ApiConsumes"), [
      builders.literal("multipart/form-data"),
    ]),
  );

  const bodyDecorator = builders.decorator(
    builders.callExpression(builders.identifier("swagger.ApiBody"), [
      builders.objectExpression([
        builders.property(
          "init",
          builders.identifier("schema"),
          builders.objectExpression([
            builders.property(
              "init",
              builders.identifier("type"),
              builders.literal("object"),
            ),
            builders.property(
              "init",
              builders.identifier("properties"),
              builders.objectExpression([
                builders.property(
                  "init",
                  builders.identifier("file"),
                  builders.objectExpression([
                    builders.property(
                      "init",
                      builders.identifier("type"),
                      builders.literal("string"),
                    ),
                    builders.property(
                      "init",
                      builders.identifier("format"),
                      builders.literal("binary"),
                    ),
                  ]),
                ),
              ]),
            ),
          ]),
        ),
      ]),
    ]),
  );

  const paramDecorator = builders.decorator(
    builders.callExpression(builders.identifier("swagger.ApiParam"), [
      builders.objectExpression([
        builders.property(
          "init",
          builders.identifier("name"),
          builders.literal("id"),
        ),
        builders.property(
          "init",
          builders.identifier("type"),
          builders.literal("string"),
        ),
        builders.property(
          "init",
          builders.identifier("required"),
          builders.literal(true),
        ),
      ]),
    ]),
  );

  const createdResponseDecorator = builders.decorator(
    builders.callExpression(builders.identifier("swagger.ApiCreatedResponse"), [
      builders.objectExpression([
        builders.property(
          "init",
          builders.identifier("type"),
          builders.identifier(pascalCase(entityId.name)),
        ),
        builders.property(
          "init",
          builders.identifier("status"),
          builders.literal("2XX"),
        ),
      ]),
    ]),
  );

  return builders.classMethod.from({
    async: true,
    key: methodId,
    params: params,
    decorators: [
      putDecorator,
      interceptorDecorator,
      consumesDecorator,
      bodyDecorator,
      paramDecorator,
      createdResponseDecorator,
      swaggerApiNotFoundResponseDecorator,
    ],
    returnType: builders.tsTypeAnnotation(
      builders.tsTypeReference(
        builders.identifier("Promise"),
        builders.tsTypeParameterInstantiation([
          builders.tsTypeReference(entityId),
        ]),
      ),
    ),
    body: builders.blockStatement([
      builders.returnStatement(
        builders.callExpression(
          builders.memberExpression(
            builders.identifier("this.service"),
            builders.identifier(`upload${pascalCase(name)}`),
          ),
          [
            builders.objectExpression([
              builders.property(
                "init",
                builders.identifier("where"),
                builders.identifier("params"),
              ),
            ]),
            builders.callExpression(builders.identifier("Object.assign"), [
              builders.identifier("file"),
              builders.objectExpression([
                builders.property(
                  "init",
                  builders.identifier("filename"),
                  builders.memberExpression(
                    builders.identifier("file"),
                    builders.identifier("originalname"),
                  ),
                ),
              ]),
            ]),
          ],
        ),
      ),
    ]),
  });
};

const getDeleteMethod = (
  params: namedTypes.Identifier[],
  methodId: namedTypes.Identifier,
  entityId: namedTypes.Identifier,
  name: string,
) => {
  const deleteDecorator = builders.decorator(
    builders.callExpression(builders.identifier("common.Delete"), [
      builders.literal(`:id/${name}`),
    ]),
  );

  const swaggerApiOkResponseDecorator = builders.decorator(
    builders.callExpression(builders.identifier("swagger.ApiOkResponse"), [
      builders.objectExpression([
        builders.property(
          "init",
          builders.identifier("type"),
          builders.identifier(pascalCase(entityId.name)),
        ),
      ]),
    ]),
  );

  return builders.classMethod.from({
    async: true,
    key: methodId,
    params: params,
    returnType: builders.tsTypeAnnotation(
      builders.tsTypeReference(
        builders.identifier("Promise"),
        builders.tsTypeParameterInstantiation([
          builders.tsTypeReference(entityId),
        ]),
      ),
    ),
    decorators: [
      deleteDecorator,
      swaggerApiOkResponseDecorator,
      swaggerApiNotFoundResponseDecorator,
    ],
    body: builders.blockStatement([
      builders.returnStatement(
        builders.callExpression(
          builders.memberExpression(
            builders.identifier("this.service"),
            builders.identifier(`delete${pascalCase(name)}`),
          ),
          [
            builders.objectExpression([
              builders.property(
                "init",
                builders.identifier("where"),
                builders.identifier("params"),
              ),
            ]),
          ],
        ),
      ),
    ]),
  });
};
