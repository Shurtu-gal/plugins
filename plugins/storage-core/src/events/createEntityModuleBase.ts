import {
  CreateEntityModuleBaseParams,
  DsgContext,
  EnumDataType,
} from "@amplication/code-gen-types";
import { addImports, importNames } from "../util/ast";
import { builders } from "ast-types";
import { camelCase } from "lodash";

export const beforeCreateEntityModuleBase = async (
  context: DsgContext,
  eventParams: CreateEntityModuleBaseParams,
) => {
  const { entities } = context;
  const { template, templateMapping, entityName } = eventParams;

  console.log("beforeCreateEntityModuleBase", entityName);

  if (
    entities
      ?.find((entity) => camelCase(entity.name) === entityName)
      ?.fields?.some((field) => field.dataType === EnumDataType.File)
  ) {
    const storageModuleId = builders.identifier("StorageModule");
    const storageModuleImport = importNames(
      [storageModuleId],
      "src/storage/storage.module",
    );

    addImports(template, [storageModuleImport]);

    const importArray = builders.arrayExpression([
      storageModuleId,
      ...templateMapping["IMPORTS_ARRAY"].elements,
    ]);

    const exportArray = builders.arrayExpression([
      storageModuleId,
      ...templateMapping["EXPORT_ARRAY"].elements,
    ]);

    templateMapping["IMPORTS_ARRAY"] = importArray;
    templateMapping["EXPORT_ARRAY"] = exportArray;
  }

  return eventParams;
};
