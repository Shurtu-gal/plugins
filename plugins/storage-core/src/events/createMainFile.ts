import { CreateMainFileParams, DsgContext } from "@amplication/code-gen-types";
import { templatesPath } from "../constants";
import { resolve } from "path";
import { readFile } from "@amplication/code-gen-utils";

export const beforCreateMainFile = async (
  context: DsgContext,
  eventParams: CreateMainFileParams,
) => {
  const templatePath = resolve(templatesPath, "main.template.ts");

  eventParams.template = await readFile(templatePath);

  return eventParams;
};
