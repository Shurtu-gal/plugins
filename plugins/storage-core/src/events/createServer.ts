import {
  CreateServerParams,
  DsgContext,
  ModuleMap,
} from "@amplication/code-gen-types";
import { serverStaticsPath } from "../constants";

export const afterCreateServer = async (
  context: DsgContext,
  eventParams: CreateServerParams,
  modules: ModuleMap,
) => {
  const { utils } = context;

  const staticFiles = await utils.importStaticModules(
    serverStaticsPath,
    context.serverDirectories.srcDirectory,
  );

  await modules.merge(staticFiles);

  return modules;
};
