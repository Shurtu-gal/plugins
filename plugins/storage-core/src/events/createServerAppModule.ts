import {
  CreateServerAppModuleParams,
  DsgContext,
  ModuleMap,
  PluginInstallation,
} from "@amplication/code-gen-types";
import { builders } from "ast-types";
import { join, resolve } from "path";
import { templatesPath } from "../constants";
import { print, readFile } from "@amplication/code-gen-utils";
import { pascalCase } from "pascal-case";
import { addImports, importNames, interpolate } from "../util/ast";

const prepareStorageModuleTemplate = async (
  pluginInstallations: PluginInstallation[],
) => {
  const storageModuleTemplatePath = resolve(
    templatesPath,
    "storage.module.template.ts",
  );

  const storageModuleTemplate = await readFile(storageModuleTemplatePath);

  // Gets name of the storage plugins that are enabled ( for this to work the plugins must have the storage prefix)
  const fileUploadPlugins = pluginInstallations
    .filter(
      (plugin) =>
        plugin.pluginId.includes("storage") &&
        plugin.enabled &&
        plugin.pluginId !== "storage-core",
    )
    .map((plugin) => plugin.pluginId.replace("storage", ""));

  // local -> LocalStorageService, s3 -> S3StorageService
  const storageServices = fileUploadPlugins.map((plugin) => {
    return builders.identifier(`${pascalCase(plugin)}StorageService`);
  });

  const templateMapping = {
    PROVIDERS_ARRAY: builders.arrayExpression(storageServices),
  };

  interpolate(storageModuleTemplate, templateMapping);

  // import { LocalStorageService } from './providers/local/local.storage.service';
  const importArray = fileUploadPlugins.map((service) =>
    importNames(
      [builders.identifier(`${pascalCase(service)}StorageService`)],
      `./providers/${service}/${service}.storage.service`,
    ),
  );

  addImports(storageModuleTemplate, importArray);

  return storageModuleTemplate;
};

export const beforeCreateServerAppModule = async (
  context: DsgContext,
  eventParams: CreateServerAppModuleParams,
) => {
  const { templateMapping, modulesFiles } = eventParams;
  const { pluginInstallations } = context;
  const moduleFile = await prepareStorageModuleTemplate(pluginInstallations);

  const storageModuleId = builders.identifier("StorageModule");

  const importArray = builders.arrayExpression([
    storageModuleId,
    ...eventParams.templateMapping["MODULES"].elements,
  ]);

  templateMapping["MODULES"] = importArray;

  const importFile = {
    code: print(moduleFile).code,
    path: join(
      context.serverDirectories.srcDirectory,
      "storage",
      "storage.module.ts",
    ),
  };

  modulesFiles.set(importFile);

  return eventParams;
};

export const afterCreateServerAppModule = async (
  context: DsgContext,
  eventParams: CreateServerAppModuleParams,
  modules: ModuleMap,
) => {
  const { pluginInstallations } = context;
  const storageModule = await prepareStorageModuleTemplate(pluginInstallations);

  const storageModuleFile = {
    code: print(storageModule).code,
    path: join(
      context.serverDirectories.srcDirectory,
      "storage",
      "storage.module.ts",
    ),
  };

  await modules.set(storageModuleFile);

  return modules;
};
