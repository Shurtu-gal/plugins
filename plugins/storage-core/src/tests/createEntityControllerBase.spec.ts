import {
  CreateEntityControllerBaseParams,
  DsgContext,
  EnumDataType,
} from "@amplication/code-gen-types";
import { print, readFile } from "@amplication/code-gen-utils";
import { mock } from "jest-mock-extended";
import { beforeCreateEntityControllerBase } from "../events";
import { join } from "path";
import { builders } from "ast-types";

describe("Test createEntityControllerBase", () => {
  let context: DsgContext;
  let eventParams: CreateEntityControllerBaseParams;

  beforeEach(() => {
    context = mock<DsgContext>();
    eventParams = mock<CreateEntityControllerBaseParams>({
      entity: {
        id: "user",
        name: "User",
        displayName: "User",
        pluralDisplayName: "Users",
        fields: [
          {
            id: "profilePicture",
            name: "profilePicture",
            displayName: "Profile Picture",
            dataType: EnumDataType.File,
            properties: {
              allowedFileTypes: ["image/jpeg", "image/png"],
              containerPath: "profile-pictures",
              maxFileSize: 10485760,
            },
            required: false,
            searchable: false,
          },
        ],
      },
      controllerBaseId: builders.identifier("UserControllerBase"),
    });
  });

  it("should add all the necessary methods", async () => {
    eventParams.template = await readFile(
      join(__dirname, "./fixtures/controller.base.template.ts"),
    );
    eventParams.templateMapping = {
      CONTROLLER_BASE: builders.identifier("UserControllerBase"),
    };

    const { template } = await beforeCreateEntityControllerBase(
      context,
      eventParams,
    );
    expect(print(template).code).toMatchSnapshot();
  });
});
