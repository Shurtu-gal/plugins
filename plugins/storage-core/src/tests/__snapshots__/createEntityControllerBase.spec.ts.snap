// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test createEntityControllerBase should add all the necessary methods 1`] = `
"import * as common from "@nestjs/common";
import * as swagger from "@nestjs/swagger";
import { isRecordNotFoundError } from "../../prisma.util";
import * as errors from "../../errors";
import { Request, Response } from "express";
import { plainToClass } from "class-transformer";
import { ApiNestedQuery } from "../../decorators/api-nested-query.decorator";
import { FileInterceptor } from "@nestjs/platform-express";

declare interface CREATE_INPUT {}
declare interface WHERE_INPUT {}
declare interface WHERE_UNIQUE_INPUT {}
declare class FIND_MANY_ARGS {
  where: WHERE_INPUT;
}
declare interface UPDATE_INPUT {}

declare const FINE_ONE_PATH: string;
declare const UPDATE_PATH: string;
declare const DELETE_PATH: string;

declare class ENTITY {}
declare interface Select {}

declare interface SERVICE {
  CREATE_FUNCTION(args: {
    data: CREATE_INPUT;
    select: Select;
  }): Promise<ENTITY>;
  FIND_MANY_FUNCTION(args: {
    where: WHERE_INPUT;
    select: Select;
  }): Promise<ENTITY[]>;
  FIND_ONE_FUNCTION(args: {
    where: WHERE_UNIQUE_INPUT;
    select: Select;
  }): Promise<ENTITY | null>;
  UPDATE_FUNCTION(args: {
    where: WHERE_UNIQUE_INPUT;
    data: UPDATE_INPUT;
    select: Select;
  }): Promise<ENTITY>;
  DELETE_FUNCTION(args: {
    where: WHERE_UNIQUE_INPUT;
    select: Select;
  }): Promise<ENTITY>;
}

declare const RESOURCE: string;
declare const ENTITY_NAME: string;
declare const CREATE_DATA_MAPPING: CREATE_INPUT;
declare const UPDATE_DATA_MAPPING: UPDATE_INPUT;
declare const SELECT: Select;

export class UserControllerBase {
  constructor(protected readonly service: SERVICE) {}
  @common.Post()
  @swagger.ApiCreatedResponse({ type: ENTITY })
  async CREATE_ENTITY_FUNCTION(
    @common.Body() data: CREATE_INPUT
  ): Promise<ENTITY> {
    return await this.service.CREATE_FUNCTION({
      data: CREATE_DATA_MAPPING,
      select: SELECT,
    });
  }

  @common.Get()
  @swagger.ApiOkResponse({ type: [ENTITY] })
  @ApiNestedQuery(FIND_MANY_ARGS)
  async FIND_MANY_ENTITY_FUNCTION(
    @common.Req() request: Request
  ): Promise<ENTITY[]> {
    const args = plainToClass(FIND_MANY_ARGS, request.query);
    return this.service.FIND_MANY_FUNCTION({
      ...args,
      select: SELECT,
    });
  }

  @common.Get(FINE_ONE_PATH)
  @swagger.ApiOkResponse({ type: ENTITY })
  @swagger.ApiNotFoundResponse({ type: errors.NotFoundException })
  async FIND_ONE_ENTITY_FUNCTION(
    @common.Param() params: WHERE_UNIQUE_INPUT
  ): Promise<ENTITY | null> {
    const result = await this.service.FIND_ONE_FUNCTION({
      where: params,
      select: SELECT,
    });
    if (result === null) {
      throw new errors.NotFoundException(
        \`No resource was found for \${JSON.stringify(params)}\`
      );
    }
    return result;
  }

  @common.Patch(UPDATE_PATH)
  @swagger.ApiOkResponse({ type: ENTITY })
  @swagger.ApiNotFoundResponse({ type: errors.NotFoundException })
  async UPDATE_ENTITY_FUNCTION(
    @common.Param() params: WHERE_UNIQUE_INPUT,
    @common.Body() data: UPDATE_INPUT
  ): Promise<ENTITY | null> {
    try {
      return await this.service.UPDATE_FUNCTION({
        where: params,
        data: UPDATE_DATA_MAPPING,
        select: SELECT,
      });
    } catch (error) {
      if (isRecordNotFoundError(error)) {
        throw new errors.NotFoundException(
          \`No resource was found for \${JSON.stringify(params)}\`
        );
      }
      throw error;
    }
  }

  @common.Delete(DELETE_PATH)
  @swagger.ApiOkResponse({ type: ENTITY })
  @swagger.ApiNotFoundResponse({ type: errors.NotFoundException })
  async DELETE_ENTITY_FUNCTION(
    @common.Param() params: WHERE_UNIQUE_INPUT
  ): Promise<ENTITY | null> {
    try {
      return await this.service.DELETE_FUNCTION({
        where: params,
        select: SELECT,
      });
    } catch (error) {
      if (isRecordNotFoundError(error)) {
        throw new errors.NotFoundException(
          \`No resource was found for \${JSON.stringify(params)}\`
        );
      }
      throw error;
    }
  }

  @common.Put(":id/profilePicture")
  @common.UseInterceptors(FileInterceptor("file"))
  @swagger.ApiConsumes("multipart/form-data")
  @swagger.ApiBody({
    schema: {
      type: "object",

      properties: {
        file: {
          type: "string",
          format: "binary"
        }
      }
    }
  })
  @swagger.ApiParam({
    name: "id",
    type: "string",
    required: true
  })
  @swagger.ApiCreatedResponse({
    type: User,
    status: "2XX"
  })
  @swagger.ApiNotFoundResponse({
    type: errors.NotFoundException
  })
  async uploadProfilePicture(
    @common.Param()
    params: UserWhereUniqueInput,
    @common.UploadedFile()
    file: Express.Multer.File
  ): Promise<User> {
    return this.service.uploadProfilePicture({
      where: params
    }, Object.assign(file, {
      filename: file.originalname
    }));
  }

  @common.Get(":id/profilePicture")
  @swagger.ApiParam({
    name: "id",
    type: "string",
    required: true
  })
  @swagger.ApiOkResponse({
    type: common.StreamableFile
  })
  @swagger.ApiNotFoundResponse({
    type: errors.NotFoundException
  })
  async downloadProfilePicture(
    @common.Param()
    params: UserWhereUniqueInput,
    @common.Res({
      passthrough: true
    })
    res: Response
  ): Promise<common.StreamableFile> {
    const result = await this.service.downloadProfilePicture({
      where: params
    });

    if (result === null) {
      throw new errors.NotFoundException(\`No resource was found for \`, JSON.stringify(params));
    }

    res.setHeader("Content-Disposition", \`attachment; filename=\${result.filename}\`);
    res.setHeader("Content-Type", result.mimetype);
    return result.stream;
  }

  @common.Delete(":id/profilePicture")
  @swagger.ApiOkResponse({
    type: User
  })
  @swagger.ApiNotFoundResponse({
    type: errors.NotFoundException
  })
  async deleteProfilePicture(
    @common.Param()
    params: UserWhereUniqueInput
  ): Promise<User> {
    return this.service.deleteProfilePicture({
      where: params
    });
  }
}
"
`;
